branchLayout/api/src/main/java/com/virtuslab/branchlayout/api/BranchLayoutEntry.java

```diff
@@ -16,27 +16,43 @@
  *  Two {@code BranchLayoutEntry} objects are equal when their names, their custom annotations
  *  and their children are ALL equal (recursively checked for children).
  */
-@AllArgsConstructor
+@AllArgsConstructor // for Lombok-generated `withChildren`
 @SuppressWarnings("interning:not.interned") // to allow for `==` comparison in Lombok-generated `withChildren` method
 @ToString
-public class BranchLayoutEntry {
+public final class BranchLayoutEntry {
   @Getter
   private final String name;
 
   @Getter
   private final @Nullable String customAnnotation;
 
+  @Getter
+  private @Nullable BranchLayoutEntry parent = null;
+
   @Getter
   @With
   private final List<BranchLayoutEntry> children;
 
+  public BranchLayoutEntry(String name, @Nullable String customAnnotation, List<BranchLayoutEntry> children) {
+    this.name = name;
+    this.customAnnotation = customAnnotation;
+    this.children = children;
+
+    // Note: since the class is final, `this` is already @Initialized at this point.
+    // We're thus safe to store `this` in other objects' fields.
+
+    for (val child : children) {
+      child.parent = this;
+    }
```

Comment:
Are you sure this logic works as expected? I would suggest adding some test / tests to make sure that `parent` is set properly.

Reply:
Whoops of course it didn't, as usually with untested code & mutable structures, there was a subtle bug around slide-outs :trollface: 

I added tests + changed the structure to completely immutable. Children are copied in c'tor to new objects with a correct parent.